### ISSUE TITLE
[DCP] DCP load for non-tensor values

### ISSUE THREAD

**Issue by ultranity** [pytorch/pytorch#125528]
### 🐛 Describe the bug

When using dcp.save or dcp.async_save as torch.save alternative and then loading state_dicts back, non-tensor values is not updated as expected, for example(edited from [async_save ](https://github.com/pytorch/pytorch/blob/release/2.3/torch/distributed/checkpoint/examples/async_checkpointing_example.py)example):
```python
import torch
import torch.distributed.checkpoint as dcp

def run(rank, world_size, resume="ckpt"):
    args = argparse.Namespace(local_rank=rank, world_size=world_size)
    state_dict = {"model": model, "optim": optim, "epoch": 0, "args":vars(args)}
    if os.path.exists(resume):
        dcp.load(state_dict, checkpoint_id=resume)
        _print(f"Resuming from epoch {state_dict['epoch']}")
    f = None
    for epoch in range(NUM_EPOCHS):
        state_dict['epoch'] = epoch
        try:
            #torch.manual_seed(epoch)
            x, y = _input()

            loss = loss_calc(model(x), y)

            _print(f"{epoch=} {loss=}")

            loss.backward()
            optim.step()
            optim.zero_grad()

            if epoch>0 and epoch % SAVE_PERIOD == 0:
                if f is not None:
                    f.result()
                f = dcp.state_dict_saver.async_save(
                    state_dict, checkpoint_id=CHECKPOINT_DIR
                )

            if epoch>0 and FAULT_PERIOD > 0 and epoch % FAULT_PERIOD == 0:
                raise InjectedException("Fault injection!")
        except InjectedException as e:
            dist.barrier()

            _print("Trainer encountered exception:")
            traceback.print_tb(e.__traceback__)

            _print("Reloading model from last checkpoint!")
            if f is not None:
                f.result()
```
1. the epoch num is expected to be updated as other state_dicts but now it will aways be 0(init value)
2. BTW Namespace cannot be saved using dcp.async_save
```
    f = dcp.async_save(
        ^^^^^^^^^^^^^^^
  File "/opt/miniforge3/envs/py/lib/python3.11/site-packages/torch/distributed/checkpoint/state_dict_saver.py", line 214, in async_save
    cpu_state_dict = _offload_state_dict_to_cpu(_stateful_to_state_dict(state_dict))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/miniforge3/envs/py/lib/python3.11/site-packages/torch/distributed/_state_dict_utils.py", line 282, in _offload_state_dict_to_cpu
    ret = _iterate_state_dict(
          ^^^^^^^^^^^^^^^^^^^^
  File "/opt/miniforge3/envs/py/lib/python3.11/site-packages/torch/distributed/_state_dict_utils.py", line 100, in _iterate_state_dict
    ret = {
          ^
  File "/opt/miniforge3/envs/py/lib/python3.11/site-packages/torch/distributed/_state_dict_utils.py", line 101, in <dictcomp>
    key: _iterate_state_dict(
         ^^^^^^^^^^^^^^^^^^^^
  File "/opt/miniforge3/envs/py/lib/python3.11/site-packages/torch/distributed/_state_dict_utils.py", line 142, in _iterate_state_dict
    raise ValueError(f"Unexpected value type {type(iter_object)}")
ValueError: Unexpected value type <class 'argparse.Namespace'>
```

### Versions

PyTorch version: 2.3.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.11.8 | packaged by conda-forge | (main, Feb 16 2024, 20:53:32) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-88-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.91
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA H100 80GB HBM3
GPU 1: NVIDIA H100 80GB HBM3
GPU 2: NVIDIA H100 80GB HBM3
GPU 3: NVIDIA H100 80GB HBM3
GPU 4: NVIDIA H100 80GB HBM3
GPU 5: NVIDIA H100 80GB HBM3
GPU 6: NVIDIA H100 80GB HBM3
GPU 7: NVIDIA H100 80GB HBM3

Nvidia driver version: 535.54.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             192
On-line CPU(s) list:                0-191
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Platinum 8468
CPU family:                         6
Model:                              143
Thread(s) per core:                 2
Core(s) per socket:                 48
Socket(s):                          2
Stepping:                           8
Frequency boost:                    enabled
CPU max MHz:                        2101.0000
CPU min MHz:                        800.0000
BogoMIPS:                           4200.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          4.5 MiB (96 instances)
L1i cache:                          3 MiB (96 instances)
L2 cache:                           192 MiB (96 instances)
L3 cache:                           210 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-47,96-143
NUMA node1 CPU(s):                  48-95,144-191
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] lion-pytorch==0.1.2
[pip3] mypy==1.8.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] pytorch-lightning==2.2.1
[pip3] pytorch-ranger==0.1.1
[pip3] torch==2.3.0
[pip3] torch-fidelity==0.3.0
[pip3] torch-optimizer==0.3.0
[pip3] torch-tb-profiler==0.4.3
[pip3] torchaudio==2.3.0
[pip3] torchmetrics==1.3.1
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.18.0
[pip3] triton==2.3.0
[conda] lion-pytorch              0.1.2                    pypi_0    pypi
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] pytorch-lightning         2.2.1                    pypi_0    pypi
[conda] pytorch-ranger            0.1.1                    pypi_0    pypi
[conda] torch                     2.3.0                    pypi_0    pypi
[conda] torch-fidelity            0.3.0                    pypi_0    pypi
[conda] torch-optimizer           0.3.0                    pypi_0    pypi
[conda] torch-tb-profiler         0.4.3                    pypi_0    pypi
[conda] torchaudio                2.3.0                    pypi_0    pypi
[conda] torchmetrics              1.3.1                    pypi_0    pypi
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchvision               0.18.0                   pypi_0    pypi
[conda] triton                    2.3.0                    pypi_0    pypi

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @LucasLLC

**Comment by ultranity** [comment]
And now there is no direct method to load full dcp saved state dict as torch.load does, while finding there is a `dcp_to_torch_save` impl. in dcp.format_utils, maybe we can make is better accessable by adding a function like `dcp_to_torch_save` but just return the state_dict? for example:
```python

def dcp_load(
    dcp_checkpoint_dir: Union[str, os.PathLike],
    no_dist=True,
):
    """
    Given a directory containing a DCP checkpoint, this function will load it into a
    state dict and return it as torch.load does.

    Args:
        dcp_checkpoint_dir: Directory containing the DCP checkpoint.

    .. warning::
        To avoid OOM, it's recommended to only run this function on a single rank.
    """
    sd: STATE_DICT_TYPE = {}

    _load_state_dict(
        sd,
        storage_reader=FileSystemReader(dcp_checkpoint_dir),
        planner=_EmptyStateDictLoadPlanner(),
        no_dist=no_dist,
    )
    return sd
```

by using this method we can check that dcp.async_save do save other values (state_dict['epoch']==20 in that case), but dcp.load is not updating the dict

**Comment by LucasLLC** [comment]
Looking into this, thanks @ultranity

**Comment by LucasLLC** [comment]
Can we replicate? I believe this was solved recently in #127617

**Comment by LucasLLC** [comment]
Sorry to clarify, we would need to pipe `type_check=False` to `_iterate_state_dict` to accept Namespace types, but the specific issue related to epoch I assume would be fixed in https://github.com/pytorch/pytorch/pull/127617

**Comment by ultranity** [comment]
> Sorry to clarify, we would need to pipe `type_check=False` to `_iterate_state_dict` to accept Namespace types, but the specific issue related to epoch I assume would be fixed in #127617

will this be solved in the coming 2.4.0 release?

**Comment by LucasLLC** [comment]
@ultranity , sorry I think we just missed the deadline for 2.4, but let me see if we can get this in before 2.4.1

**Comment by ultranity** [comment]
> @ultranity , sorry I think we just missed the deadline for 2.4, but let me see if we can get this in before 2.4.1

Any updates?

**Comment by fegin** [comment]
@LucasLLC  Did we get the diff in 2.4.1, if not should we aim for 2.4.2?

**Comment by fegin** [comment]
cc., @pradeepfn

**Comment by pradeepfn** [comment]
cc @daulet-askarov

**Comment by daulet-askarov** [comment]
> > Sorry to clarify, we would need to pipe `type_check=False` to `_iterate_state_dict` to accept Namespace types, but the specific issue related to epoch I assume would be fixed in #127617
> 
> cc: @LucasLLC do you know why we set `type_check=False` in `async_save`, the default is `type_check=True`.

> ```python
>     args = argparse.Namespace(local_rank=rank, world_size=world_size)
>     state_dict = {"model": model, "optim": optim, "epoch": 0, "args":vars(args)}
> ```

I think Lucas meant it in the context of the OP's desire to put Namespace type object into state dictionary, and asynchronously save it. See these two lines in the issue description.
    args = argparse.Namespace(local_rank=rank, world_size=world_size)
    state_dict = {"model": model, "optim": optim, "epoch": 0, "args":vars(args)}
type_check=True would enforce state_dict values to be only of these types:
https://github.com/pytorch/pytorch/blob/main/torch/distributed/_state_dict_utils.py#L114-L116
otherwise ValueError is raised:
https://github.com/pytorch/pytorch/blob/main/torch/distributed/_state_dict_utils.py#L185-L188

**Comment by ankitageorge** [comment]
closing this out, as both issues should be solved now
